`torch.pow` Add type promotion support and fix issue with __rpow__

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32582 `torch.pow` Add type promotion support and fix issue with __rpow__**
* #32581 Simplify precision in type promotion tests.
* #32538 Use allclose and relative tolerance in unit testing

Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

Also out-of-place int_tensor ** float => float_tensor, whereas in 1.4 the result dtype did not change.

nairbv committed:
Fix issues with `pow` type promotion

Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 384db39c:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=384db39cacfa35bf138a7d7398ebd7686e2fa6d8).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4332678) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90368971) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90368971))
<details>
<summary>
<code>Jan 24 17:01:42 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff</code>
</summary>

```
Jan 24 17:01:42 File to patch:  
Jan 24 17:01:42 Skip this patch? [y]  
Jan 24 17:01:42 4 out of 4 hunks ignored 
Jan 24 17:01:42 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 24 17:01:42 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 24 17:01:42 Traceback (most recent call last): 
Jan 24 17:01:42   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 24 17:01:42     patch_repo(args) 
Jan 24 17:01:42   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 24 17:01:42     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 24 17:01:42 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 24 17:01:42 + cleanup 
Jan 24 17:01:42 + retcode=1 
Jan 24 17:01:42 + set +x 
Jan 24 17:01:42 =================== sccache compilation log =================== 
Jan 24 17:01:42 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 24 17:01:42 Compile requests                 1 
Jan 24 17:01:42 Compile requests executed        0 
Jan 24 17:01:42 Cache hits                       0 
Jan 24 17:01:42 Cache misses                     0 
Jan 24 17:01:42 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv commented on aten/src/ATen/native/cpu/PowKernel.cpp:
```diff
@@ -148,89 +148,13 @@ void pow_tensor_scalar_kernel(TensorIterator& iter, Scalar exp_scalar) {
-    // Integral types do not allow AVX2 vector optimizations for pow/sqrt/rsqrt.
-    // Trying to implement pow/sqrt/rsqrt as loop in vec256_int.h does not allow
-    // powering integral tensor to float exponent. That's why we need this code
-    // duplication:
```
@pbelevich with type promotion the base and exp should get promoted to the same type (both scalar_t) so integral tensor to float exponent doesn't make sense to me here. I had some build failures in avx that I'm trying to fix though so maybe I'm misunderstanding something here?

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "Fix issues with `pow` type promotion"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"


Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
     * tensors to negative powers were giving incorrect results (1 instead
        of 0 or error)
     * Behavior wasn't consistent between cuda/cpu
     * large_value ** 1 in some cases gave a result not equal
        to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

[ghstack-poisoned]

ngimel commented:
Re: BC breaking change, do we generally want our type promotion rules to match numpy? Because numpy allows exponentiating integer ndarrays with float exponent, producing float ndarray (which is still different from 1.4 behavior)
```
In [4]: a=np.arange(10, dtype=np.int32)                                                                                               

In [5]: a                                                                                                                             
Out[5]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=int32)

In [6]: np.power(a,0.5)                                                                                                               
Out[6]: 
array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])
```

nairbv commented:
> Re: BC breaking change, do we generally want our type promotion rules to match numpy? 

Yes, generally, though our casting/promotion rules aren't perfectly identical.

> Because numpy allows exponentiating integer ndarrays with float exponent, producing float ndarray (which is still different from 1.4 behavior)

This does too (and is BC-breaking in multiple ways, I'll clear that up in the description). The error message is only for in-place `pow_`, which is equivalent to np:

```
a=np.arange(10, dtype=np.int32)
np.power(a, 0.5, out=a)
TypeError: ufunc 'power' output (typecode 'd') could not be coerced to provided output parameter (typecode 'i') according to the casting rule ''same_kind''
```

nairbv commented:
replaced by:
https://github.com/pytorch/pytorch/pull/37098
